src/meta/meta.go

```diff
@@ -0,0 +1,166 @@
+package meta
+
+import (
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	// NamespaceSeparator used by the K8s informers library to create indices
+	// that are composed as namespace/name
+	NamespaceSeparator = "/"
+	IndexIP            = "IP"
+)
+
+var slog = logrus.WithFields(logrus.Fields{
+	"component": fmt.Sprintf("%T", Informers{}),
+})
+
+type Informers struct {
+	informerFactory informers.SharedInformerFactory
+	pods            cache.SharedIndexInformer
+	services        cache.SharedIndexInformer
+	replicaSet      cache.SharedIndexInformer
+}
+
+func NewInformers(client kubernetes.Interface) Informers {
+	// TODO: configure resync time
+	factory := informers.NewSharedInformerFactory(client, 1*time.Hour)
+	pods := factory.Core().V1().Pods().Informer()
+	if err := pods.AddIndexers(map[string]cache.IndexFunc{
+		IndexIP: func(obj interface{}) ([]string, error) {
+			pod := obj.(*corev1.Pod)
+			ips := make([]string, 0, len(pod.Status.PodIPs))
+			for _, ip := range pod.Status.PodIPs {
+				// ignoring host-networked Pod IPs
+				if ip.IP != pod.Status.HostIP {
+					ips = append(ips, ip.IP)
+				}
+			}
+			return ips, nil
+		},
+	}); err != nil {
+		// this should never happen, as it only returns error if the informer has
+		// been alrady started
+		panic(err)
+	}
+	services := factory.Core().V1().Services().Informer()
+	if err := services.AddIndexers(map[string]cache.IndexFunc{
+		IndexIP: func(obj interface{}) ([]string, error) {
+			spec := obj.(*corev1.Service).Spec
+			if spec.ClusterIP == corev1.ClusterIPNone {
+				return []string{}, nil
+			}
+			return spec.ClusterIPs, nil
+		},
+	}); err != nil {
+		panic(err)
+	}
+	return Informers{
+		informerFactory: factory,
+		pods:            pods,
+		services:        services,
+		replicaSet:      factory.Apps().V1().ReplicaSets().Informer(),
+	}
+}
+
+func (s *Informers) Start(stopCh <-chan struct{}) error {
+	s.informerFactory.Start(stopCh)
+	return nil
+}
+
+func (s *Informers) WaitForCacheSync(stopCh <-chan struct{}) {
+	s.informerFactory.WaitForCacheSync(stopCh)
+}
+
+func (s *Informers) PodByIP(ip string) (*corev1.Pod, bool) {
+	item, err := s.pods.GetIndexer().ByIndex(IndexIP, ip)
+	if err != nil {
+		// should never happen as long as we provide the correct index function
+		// otherwise it's a bug in our code
+		panic(err)
+	}
+	// our provided indexers only return a key, so it's safe to assume 0<=len()<=1
+	if len(item) == 0 {
+		// not found
+		return nil, false
+	}
+	// since we are excluding host-networked pods, the relation IP:Pod is 1:1.
```

Comment:
I think I still prefer to log a warning if there's more than 1 item, there might be cases that we don't know yet a for which it would be useful to not silent this (like I didn't know about the host-networked pods not long ago)